lib/src/device_authorization_grant.dart

```diff
@@ -0,0 +1,401 @@
+
+  /// Callback to be invoked whenever the credentials are refreshed.
+  ///
+  /// This will be passed as-is to the constructed [Client].
+  final CredentialsRefreshedCallback? _onCredentialsRefreshed;
+
+  /// Whether to use HTTP Basic authentication for authorizing the client.
+  final bool _basicAuth;
+
+  /// A [String] used to separate scopes; defaults to `" "`.
+  final String _delimiter;
+
+  /// The HTTP client used to make HTTP requests.
+  final http.Client _httpClient;
+
+  /// The scopes that the client is requesting access to.
+  List<String>? _scopes;
+
+  /// The current state of the grant object.
+  _State _state = _State.initial;
+
+  /// The deviceId
+  String? _deviceCode;
+
+  /// Creates a new grant.
+  ///
+  /// If [basicAuth] is `true` (the default), the client credentials are sent to
+  /// the server using using HTTP Basic authentication as defined in [RFC 2617].
+  /// Otherwise, they're included in the request body. Note that the latter form
+  /// is not recommended by the OAuth 2.0 spec, and should only be used if the
+  /// server doesn't support Basic authentication.
+  ///
+  /// [RFC 2617]: https://tools.ietf.org/html/rfc2617
+  ///
+  /// [httpClient] is used for all HTTP requests made by this grant, as well as
+  /// those of the [Client] is constructs.
+  ///
+  /// [onCredentialsRefreshed] will be called by the constructed [Client]
+  /// whenever the credentials are refreshed.
+  ///
+  /// The scope strings will be separated by the provided [delimiter]. This
+  /// defaults to `" "`, the OAuth2 standard, but some APIs (such as Facebook's)
+  /// use non-standard delimiters.
+  ///
+  /// By default, this follows the OAuth2 spec and requires the server's
+  /// responses to be in JSON format. However, some servers return non-standard
+  /// response formats, which can be parsed using the [getParameters] function.
+  ///
+  /// This function is passed the `Content-Type` header of the response as well
+  /// as its body as a UTF-8-decoded string. It should return a map in the same
+  /// format as the [standard JSON response][].
+  ///
+  /// [standard JSON response]: https://tools.ietf.org/html/rfc6749#section-5.1
+  DeviceAuthorizationGrant(
+    this.identifier,
+    this.deviceEndpoint,
+    this.tokenEndpoint, {
+    this.secret,
+    String? delimiter,
+    bool basicAuth = true,
+    http.Client? httpClient,
+    CredentialsRefreshedCallback? onCredentialsRefreshed,
+    Map<String, dynamic> Function(MediaType? contentType, String body)?
+        getParameters,
+  })  : _basicAuth = basicAuth,
+        _httpClient = httpClient ?? http.Client(),
+        _delimiter = delimiter ?? ' ',
+        _getParameters = getParameters ?? parseJsonParameters,
+        _onCredentialsRefreshed = onCredentialsRefreshed;
+
+  /// starts the Flow and returns [DeviceProperties]
+  /// [useSecret] whether to use the clientSecret for the token-request or not
+  Future<DeviceProperties> getDeviceCode(
```

Comment:
We throw `Exception` when users are expected to sometimes want to handle them... Certainly we should document, that this throw `FormatException` if the response from the server invalid.